### DISCUSSION TITLE
how to implement editor inspector

### DISCUSSION THREAD

**Discussion by MasonAlice** [vsg-dev/VulkanSceneGraph#1348] [General]
Recently, I’ve been researching how to use VSG to develop an editor for a flight simulation visual engine. It seems that VSG doesn't have reflection capabilities. So how can I implement an inspector for the editor using VSG?

**Comment by robertosfield** [comment]
The VSG doesn't have general purpose reflection/introspection capability built-in, but it does have serialization support built-in to all scene graph objects.  I have had it mind that these serialization support could be leveraged for the purpose of reflection/introspection by writing a custom vsg::Input/vsg::Output.  

I haven't tried experiments with it yet, but I think some limited capabilities will be possible with the present implementation, but have had it in mind that the VSG's serialization scheme might be able to evolve to fill in the gaps of functionality required.

**Comment by AnyOldName3** [comment]
You wouldn't even necessarily want to expose VSG internals directly to an editor - a user might find it easier to place a plane rather than the scenegraph structure needed to render a plane, so you might be better served with a plane class with some properties that are exposed to the editor, and some internal state that's kept hidden, and could update the internal state as appropriate when the public properties are modified.

**Reply by MasonAlice** [reply]
Thank you for your reply. I will keep the core nodes of VSG unchanged and use the RTTR library for reflection on the custom extended nodes.
